app/pages/MainPage/Main.js

```diff
@@ -54,66 +49,47 @@ class Main extends React.Component {
       dataSource: new ListView.DataSource({
         rowHasChanged: (row1, row2) => row1 !== row2
       }),
-      typeIds: [],
       typeList: {}
     };
   }
 
   componentDidMount() {
-    const { readActions } = this.props;
-    DeviceEventEmitter.addListener('changeCategory', (typeIds) => {
-      typeIds.forEach((typeId) => {
-        readActions.requestArticleList(false, true, typeId);
-        pages.push(1);
-      });
+    this.retrieveArticleLists(this.props.selectedCategoryIds);
+
+    store.get('typeList').then((typeList) => {
       this.setState({
-        typeIds
-      });
-    });
-    InteractionManager.runAfterInteractions(() => {
-      store.get('typeIds').then((typeIds) => {
-        if (!typeIds) {
-          return;
-        }
-        typeIds.forEach((typeId) => {
-          readActions.requestArticleList(false, true, typeId);
-          pages.push(1);
-        });
-        store.get('typeList').then(typeList =>
-          this.setState({
-            typeIds,
-            typeList
-          })
-        );
+        typeList
       });
     });
   }
 
   componentWillReceiveProps(nextProps) {
     const { read } = this.props;
+    this.retrieveArticleListsWhenNeeded(
```

Comment:
and there are bugs, unfortunately, 
one is the main page can have more than 5 categories after selecting more than 5 topics and press the main page since the selectedCategoryIds are dynamically updated when you select;
another one I encountered when doing selecting and unselecting back and forth. the main page will be blank until you manually swipe left or right like below, also the highlight is wrong: 
<img width="374" alt="screen shot 2017-08-16 at 1 41 24 pm" src="https://user-images.githubusercontent.com/2666544/29347477-44f32182-8290-11e7-8262-fc0c3f891264.png">

I am not sure why the 2nd one happens. could you pls have a look when available. many thanks: